### PR TITLE
chore: Support for react-native-macos

### PR DESCRIPTION
**Pull request by asabhaney**
This PR makes some minor changes to be able to support this library on native MacOS apps if using Microsoft's [react-native-macos](https://github.com/microsoft/react-native-macos) library.

Note that I made a change to `package.json` that is unrelated to this PR, as it seems there were some changes made in #265 that breaks the installation of this lib (not quite sure why, could be that `yarn pod-install` is not occurring in the `ios` directory).

### CODE REVIEW

**Comment by mikehardy** [comment]
Oh, I'm excited to see this, thank you! I was waiting to do anything react-native-macos-related until they got 0.66 out but I understand it was released recently, so we have a modern version to try. Give me a little bit of time to review this and make sure it's good, but thanks very much for posting it, I'm excited to collaborate and get it merged

File: ios/RNAppleAuthentication/RNAppleAuthButtonViewManager.m [hunk 1]
```diff
@@ -18,6 +18,12 @@
 #import <React/RCTViewManager.h>
 #import "RNAppleAuthButtonView.h"
 
+#if TARGET_OS_OSX
+@compatibility_alias RNCPView NSView;
```
Comment (mikehardy) [inline ios/RNAppleAuthentication/RNAppleAuthButtonViewManager.m:22]: nice trick, going to use that in other areas as needed

File: ios/RNAppleAuthentication/RNAppleAuthButtonViewManager.m [hunk 2]
```diff
@@ -70,9 +76,9 @@ @implementation RNAppleAuthButtonViewManagerWhiteSignUp
   view.cornerRadius = [json floatValue];
 }
 
-- (UIView *)view {
+- (RNCPView *)view {
     ASAuthorizationAppleIDButtonType type = ASAuthorizationAppleIDButtonTypeDefault;
-    if (@available(iOS 13.2, *)) {
+    if (@available(iOS 13.2, macOS 10.15, *)) {
```
Comment (mikehardy) [inline ios/RNAppleAuthentication/RNAppleAuthButtonViewManager.m:81]: This is really subtle, but the added the signup type after initial release, macOS only 10.15.1+ https://developer.apple.com/documentation/authenticationservices/asauthorizationappleidbuttontype/asauthorizationappleidbuttontypesignup?language=objc

```suggestion
    if (@available(iOS 13.2, macOS 10.15.1, *)) {
```

File: ios/RNAppleAuthentication/RNAppleAuthButtonViewManager.m [hunk 3]
```diff
@@ -130,9 +136,9 @@ @implementation RNAppleAuthButtonViewManagerWhiteOutlineSignUp
     view.cornerRadius = [json floatValue];
   }
 
-  - (UIView *)view {
+  - (RNCPView *)view {
       ASAuthorizationAppleIDButtonType type = ASAuthorizationAppleIDButtonTypeDefault;
-      if (@available(iOS 13.2, *)) {
+      if (@available(iOS 13.2, macOS 10.15, *)) {
```
Comment (mikehardy) [inline ios/RNAppleAuthentication/RNAppleAuthButtonViewManager.m:141]: signup added after initial release per above

```suggestion
      if (@available(iOS 13.2, macOS 10.15.1, *)) {
```

File: ios/RNAppleAuthentication/RNAppleAuthButtonViewManager.m [hunk 4]
```diff
@@ -191,12 +197,11 @@ @implementation RNAppleAuthButtonViewManagerBlackSignUp
   RCT_CUSTOM_VIEW_PROPERTY(cornerRadius, NSNumber *, RNAppleAuthButtonView) {
     view.cornerRadius = [json floatValue];
   }
-  - (UIView *)view {
+  - (RNCPView *)view {
       ASAuthorizationAppleIDButtonType type = ASAuthorizationAppleIDButtonTypeDefault;
-      if (@available(iOS 13.2, *)) {
+      if (@available(iOS 13.2, macOS 10.15, *)) {
```
Comment (mikehardy) [inline ios/RNAppleAuthentication/RNAppleAuthButtonViewManager.m:202]:
```suggestion
      if (@available(iOS 13.2, macOS 10.15.1, *)) {
```

File: ios/RNAppleAuthentication/RNAppleAuthModule.m [hunk 5]
```diff
@@ -40,8 +40,8 @@ + (BOOL)requiresMainQueueSetup {
 
 - (NSDictionary *)constantsToExport {
   return @{
-      @"isSupported": @available(iOS 13.0, *) ? @(YES) : @(NO),
-      @"isSignUpButtonSupported": @available(iOS 13.2, *) ? @(YES) : @(NO),
+      @"isSupported": @available(iOS 13.0, macOS 10.15, *) ? @(YES) : @(NO),
+      @"isSignUpButtonSupported": @available(iOS 13.2, macOS 10.15, *) ? @(YES) : @(NO),
```
Comment (mikehardy) [inline ios/RNAppleAuthentication/RNAppleAuthModule.m:44]: Who is even on 10.15.0? Who knows, but it's the docs :-)

```suggestion
      @"isSignUpButtonSupported": @available(iOS 13.2, macOS 10.15.1, *) ? @(YES) : @(NO),
```

File: package.json [hunk 6]
```diff
@@ -13,13 +13,12 @@
     "build:docs": "typedoc --excludeExternals --out typedocs --plugin typedoc-plugin-markdown --hideBreadcrumbs --entryPoints lib/index.d.ts",
     "precommit": "yarn prepare && yarn build:all",
     "prepare": "yarn build && yarn tsc:compile && yarn example:prepare",
-    "example:prepare": "cd example && yarn && yarn pod-install",
+    "example:prepare": "cd example && yarn && cd ios && (pod install || true)",
```
Comment (mikehardy) [inline package.json:16]: I did this, I don't want to revert it. `pod-install` is definitely my preferred way of installing pods these days.
It's resilient to most of the common causes of users having pod install failures and auto-solves most of them.

What's the root cause of failure here? We need to diagnose it and fix that so we can still use pod-install

File: package.json [hunk 7]
```diff
@@ -13,13 +13,12 @@
...
     "example:build:android": "cd example/android && ./gradlew assembleDebug",
     "example:build:ios": "cd example && (xcodebuild -workspace ios/testing.xcworkspace -scheme testing -configuration Debug -sdk iphonesimulator -derivedDataPath ios/build -UseModernBuildSystem=YES || true)",
     "example:start": "cd example && yarn start",
-    "example:ios": "cd example && yarn run:ios",
-    "example:ios13": "cd example && yarn run:ios13",
```
Comment (mikehardy) [inline package.json:21]: why did you remove that one? I use that target all the time, the ios13 simulator is the only simulator that works, and I frequently don't have a real device around to test when checking things in this repo

File: package.json [hunk 8]
```diff
@@ -13,13 +13,12 @@
     "build:docs": "typedoc --excludeExternals --out typedocs --plugin typedoc-plugin-markdown --hideBreadcrumbs --entryPoints lib/index.d.ts",
     "precommit": "yarn prepare && yarn build:all",
     "prepare": "yarn build && yarn tsc:compile && yarn example:prepare",
-    "example:prepare": "cd example && yarn && yarn pod-install",
+    "example:prepare": "cd example && yarn && cd ios && (pod install || true)",
     "example:build:android": "cd example/android && ./gradlew assembleDebug",
     "example:build:ios": "cd example && (xcodebuild -workspace ios/testing.xcworkspace -scheme testing -configuration Debug -sdk iphonesimulator -derivedDataPath ios/build -UseModernBuildSystem=YES || true)",
     "example:start": "cd example && yarn start",
-    "example:ios": "cd example && yarn run:ios",
```
Comment (mikehardy) [inline package.json:20]: I don't understand why this line needed to change

File: package.json [hunk 9]
```diff
@@ -13,13 +13,12 @@
...
-    "example:ios13": "cd example && yarn run:ios13",
-    "example:android": "cd example && yarn run:android",
+    "example:ios": "cd example && yarn ios",
+    "example:android": "cd example && yarn android",
```
Comment (mikehardy) [inline package.json:21]: I don't understand why this line needed to change

Comment (mikehardy) [reply package.json:21]: I don't see a target for running the example with macos?

**Review by mikehardy — CHANGES_REQUESTED** [review]
mostly minor stuff, this looks like almost exactly what I'd expect for a macOS port, with the exception that your compatibility alias is much cleaner than throwing IFDEFs around everywhere, nice

**Comment by asabhaney** [comment]
Thanks for the quick review @mikehardy - I should be able to address all those issues pretty quickly. Good catch on the 10.15.1+ compatibility!

**Comment by mikehardy** [comment]
Well I did say I was excited to see the PR ;-), I still need to pull it locally and test it though, that'll be the thing that takes time, pulling the PR, making sure the example works etc - we don't have CI for it actually working. Maybe also adding CI to make sure the macos target builds (otherwise you know it'll break at some point...). But once the code is good (and it looked close) that's the big first step. Cheers

**Review by mikehardy — APPROVED** [review]
LGTM - now comes the patience part, I want to pull this and check it locally and see if I can get the example working (or *a* example working, perhaps by shifting to react-native-test-app) and that could take some time.
I can't see any reason why the code here (minus some example changes) wouldn't go in though, stated differently, you should be able to confidently use patch-package to apply these changes locally and I'll commit to notification of some sort (and collaboration) if that becomes not true in the future. And then the next release should just retire the patch

Thanks again for the port!

**Comment by mikehardy** [comment]
To be more concrete about that I'll actually merge so you can even point to a commit hash if desired, and make any necessary changes as a follow on prior to release

**Comment by asabhaney** [comment]
Great, thanks @mikehardy! I actually did start working on the Mac example (https://github.com/mother/react-native-apple-authentication/tree/macos-example), which does seems to be working, but ironically I must have broke the iOS example (though I have both iOS & macOS versions working in my own project). I'll circle back to this in the next few days to see if I can fix the example stuff.

**Comment by mikehardy** [comment]
Hold off on the example just a bit, but thanks for giving it a shot. I'm collaborating with someone in a PR to convert to react-native-test-app in another repo and if it's a winner we may be able to get out if the business of maintaining a test app completely

I maintain so many repos that's a huge 2022 goal of mine so I'm going to pursue this now with this repo as one of the reasons to prioritize it

**Comment by asabhaney** [comment]
Thanks @mikehardy! Let me know if you'd like any help with the example or tests for macOS :)

**Pull request merged**